### ISSUE TITLE
Is it possible for commands to get a numeric argument?

### ISSUE THREAD

**Issue by rdrg109** [atlas-engineer/nyxt#1756]
This is not a bug report. It is a question.

# Table of Contents

1.  [Prologue](#org3f4ef86)
    1.  [Numeric arguments in Emacs](#org7a65101)
    2.  [Relation of `C-u` and numeric arguments](#orgaf4290c)
    3.  [More information on this topic](#org07c7e29)
2.  [The context](#org2e16dc8)
3.  [The question](#orgfb07a4e)
4.  [Additional context](#org7a92ed3)

<a id="org3f4ef86"></a>

# Prologue

<a id="org7a65101"></a>

## Numeric arguments in Emacs

In Emacs, it is possible to provide a function with a numeric argument so that the behavior of a function is changed.

Some functions consider the numeric argument as the number of times the function need to be repeated. Here are some examples:

-   When `open-line` is provided with a positive numeric argument, let's call it N, N newlines are inserted in point.
-   When `kill-line` is provided with a positive numeric argument, let's call it N, N lines below the current point are killed.

Other functions use the numeric argument as a modifier of the function. Here are some examples:

-   `buffer-menu` show a list of buffers. When the function is called with a numeric argument equal to 4, only those buffers visiting files are listed.
-   `list-directory` executes `ls` in a given directory. When the function is called with a numeric argument equal to 4, `ls -l` is executed instead.
-   `where-is` shows which keys are mapped to a given function. When the function is called with a numeric argument equal to 4, the information is inserted in the current point.
-   `goto-line` prompts for a line numbering and moves the cursor to that specific line. When the function is provided with a numeric argument, the cursor is moved to the line whose numbering is the numeric argument (e.g. C-u 30 M-g M-g moves the cursor to the 30th line)

<a id="orgaf4290c"></a>

## Relation of `C-u` and numeric arguments

`C-u` provides functions with a numeric argument equal to 4, `C-u C-u <<function call>>` with 16, `C-u C-u C-u` with 64, and so on.

<a id="org07c7e29"></a>

## More information on this topic

More information on this topic can be found in `(info "(emacs) Arguments")`. For those who don't have Emacs running, you can find that Info node in [this](https://www.gnu.org/software/emacs/manual/html_node/emacs/Arguments.html) URL.

<a id="org2e16dc8"></a>

# The context

I have created the following command for creating an Org Mode headline of the currently visited URL.

```lisp
(define-command-global my/org-current-site-create-headline-and-copy ()
  "Create a headline for the current site and insert it to the
clipboard.

The title and the URL of the current buffer is used as the title of
the headline and the LINK property of that headline, respectively.

If the current buffer doesn't have a URL, then nothing is done.

For example, calling this command in the Emacs official site will
insert the following in the clipboard

* GNU Emacs - GNU Project
:PROPERTIES:
:LINK: https://www.gnu.org/software/emacs
:END:
"
  (let ((url (render-url (url (current-buffer))))
        (title (title (current-buffer)))
        data)
    (cond ((equal url "")
           (echo-warning "The current buffer doesn't have a URL"))
          (t
           (when (equal title "")
             (setq title "Unknown title"))
           (setq data (format nil "~{~A~^~%~}~%"
                              (list
                               ;; TODO: With a N prefix argument,
                               ;; repeat the "*" character N times.
                               (str:concat "* " title)
                               ":PROPERTIES:"
                               (str:concat ":LINK: " url)
                               ":END:")))
           (echo "Copied to clipboard ~A" data)
           (copy-to-clipboard data)))))
```

As you can read in the `TODO` comment, I want to provide the command with a numeric argument, so that as many asterisks as the numeric argument are used for the created Org Mode headline.

<a id="orgfb07a4e"></a>

# The question

Do Nyxt commands currently support getting the numeric argument? If not, are there any plans on making Nyxt support this?

<a id="org7a92ed3"></a>

# Additional context

I've inspected the source code and I've noticed that `C-u` is included in the definition of mappings (see below), and this makes me think that Nyxt don't currently support providing commands with the numeric argument.

I asked on #nyxt for this, and I was told that a current workaround would be to map 1 through 9 to different calls of the command (i.e. `C-u 1 {command}` `C-u 2 {command}`, &#x2026;, `C-u 9 {command}`. While this solution might suit my needs for this scenario, I thought that some day I will want to do something more complex and it would be difficult to map all the numbers, so I decided to create this issue. In addition to that, I would rather not having to map all the possible numbers.

**Comment by aartaka** [comment]
Ok, so several use-cases there:
- The so-called "universal prefix" of four: we won't have this. We have prompt-buffer actions and separate commands for that.
- Repeating action: that's possible, see #1601.
- Providing numeric argument to commands: that's quirky. You can possibly do that, but that'll end up being a per-command hack. Maybe refactor this to separate actions? Taking your example, maybe have `insert-as-*`, `insert-as-**` and `insert-as-***` actions?
  - I'm conscious that it's not as optimal as numeric prefix key. @jmercouris, @Ambrevar, @pdelfino, any ideas there?

**Comment by gregoryhugaerts** [comment]
Emacs seems to do it with a global variable. Wouldn't it be fairly easy to define such a variable, make command that sets or multiplies that variable and have commands read the variable if they want to react to it?

**Comment by Ambrevar** [comment]
Yes that would be an easy way to implement it, although globality
would preferably be avoided here.

First, I'd like to have a good use case before implementing anything.
We don't have many commands where numeric arguments make sense.  Any suggestion?

**Comment by gregoryhugaerts** [comment]
There's a few places where it could be handy:

* using numeric argument itself
  - Repeating commands, but there's repeat-times for that
  - Repeating certain parts of commands. I don't think there's any in nyxt itself, but could be nice to have for user provided ones.
*  Using the presence of a numeric argument as a toggle as is done for some commands in emacs
    - for similar but distinct commands like load-url and load-url-in-new-window. You could bind one and use universal command to call the other. This could also be achieved by binding the second command as C-u key
    -  As a toggle to enable less frequently used subparts of a command. Guessing there's none in non customized nyxt either.
  

Maybe universal argument would be better implemented in a users init.lisp file?

**Comment by rdrg109** [comment]
@Ambrevar As you mentioned, this feature isn't probably useful for the existing commands, but I think that having it implemented can attract some Emacs users since that way Nyxt would have more similarity with the Emacs environment. In addition to that, it would give more opportunities for users to create functions that suit their needs.
